### PR TITLE
Bratseth/unlimited cpu in dev

### PR DESCRIPTION
**Pull request by bratseth**
@freva please review

### CODE REVIEW

**Comment by bratseth** [comment]
In particular please verify my ContainerResources.from javadoc.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java [hunk 1]
```diff
@@ -36,11 +41,20 @@
             throw new IllegalArgumentException("memoryBytes must be a positive integer or 0 for unlimited, was " + memoryBytes);
     }
 
+    /**
+     * Create container resources from required fields.
+     *
+     * @param cpus the amount of vcpu that should be exclusively available to this container. This is a hard limit:
+     *             More than this amlunt will never be available. To allow an unlimited amount use 0.
+     * @param cpuCores the amount of vcpu that should ideally be allocated to this container if there are no other
```
Comment (freva) [inline docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: This is a more of a guarantee of lower bound of vcpus: The container is allocated a number of shares proportional to its vcpus. These shares are only used if _host_ CPU is near fully allocated (by allocating CPU time proportional to the number of shares held by each container). Therefore, a container can go above this value, but another container cannot  starve this container.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java [hunk 2]
```diff
@@ -36,11 +41,20 @@
...
+     *                 constraints on resources. To allow an unlimited amount use 0.
+     * @param memoryGb the exact amount of memory that must be available to this container.
```
Comment (freva) [inline docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: This is the same as `cpus`: it's a hard upper bound for the container and not a guarantee to it.

**Review by freva — COMMENTED** [review]
I think we should still set `cpuCores` in `ContainerResources` as that only prevents starvation of other containers, after that, the container is allowed to use as much of CPU as is left on the entire host. 

But the first argument is already "optional" in that it is controlled by a feature flag whose default value is 0, so this PR is not needed?

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java [hunk 3]
```diff
@@ -36,11 +41,20 @@
             throw new IllegalArgumentException("memoryBytes must be a positive integer or 0 for unlimited, was " + memoryBytes);
     }
 
+    /**
+     * Create container resources from required fields.
+     *
+     * @param cpus the amount of vcpu that should be exclusively available to this container. This is a hard limit:
```
[the remark this answers is not in the document]
Comment (bratseth) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: Yes, it's only a guarantee given our allocation policy, that's why I say "should". I'll make that explicit.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java [hunk 4]
```diff
@@ -36,11 +41,20 @@
...
+     *             More than this amlunt will never be available. To allow an unlimited amount use 0.
```
[the remark this answers is not in the document]
Comment (bratseth) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: Thanks.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java (hunk 1 above)
Comment (bratseth) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: That makes sense, thanks. I'll rephrase.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java (hunk 2 above)
Comment (bratseth) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: Same as the above. We actually guarantee it, and must do so always.

**Comment by bratseth** [comment]
> I think we should still set cpuCores in ContainerResources as that only prevents starvation 
> of other containers, after that, the container is allowed to use as much of CPU as is left on the entire host.

Ok, I'll do that.

> But the first argument is already "optional" in that it is controlled by a feature flag whose default value is 0, so this PR is not needed?

That feature flag should be turned on globally (and eventually be removed), but then dev should still work like this.

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (freva) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: I understand, but this is javadoc for docker-api which should be independent on what the allocation policy is.

**Review by freva — APPROVED** [review]
:100:

File: docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bratseth) [reply docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/ContainerResources.java]: Imho this method is the interface between those two perspectives and I added this javadoc to make that mapping clearer to myself.

**Pull request merged**